File: corpan/plugins/tauri-plugin-admob/android/src/main/java/com/corpora/admob/AdmobPlugin.kt
package com.corpora.admob

import android.app.Activity
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.FullScreenContentCallback
import com.google.android.gms.ads.LoadAdError
import com.google.android.gms.ads.MobileAds
import com.google.android.gms.ads.interstitial.InterstitialAd
import com.google.android.gms.ads.interstitial.InterstitialAdLoadCallback
import com.google.android.gms.ads.rewarded.RewardedAd
import com.google.android.gms.ads.rewarded.RewardedAdLoadCallback
import app.tauri.annotation.Command
import app.tauri.annotation.InvokeArg
import app.tauri.annotation.TauriPlugin
import app.tauri.plugin.Invoke
import app.tauri.plugin.JSObject
import app.tauri.plugin.Plugin

@InvokeArg
internal class AdUnitArgs {
    var adUnitId: String? = null
}

@TauriPlugin
class AdmobPlugin(private val activity: Activity) : Plugin(activity) {

    private var interstitialAd: InterstitialAd? = null
    private var rewardedAd: RewardedAd? = null
    private var isInitialized = false

    // Google's official test ad unit IDs
    private val testInterstitialId = "ca-app-pub-3940256099942544/1033173712"
    private val testRewardedId = "ca-app-pub-3940256099942544/5224354917"

    @Command
    fun initAdmob(invoke: Invoke) {
        if (isInitialized) {
            invoke.resolve()
            return
        }

        activity.runOnUiThread {
            MobileAds.initialize(activity) { status ->
                println("[ADMOB] SDK initialized: ${status.adapterStatusMap}")
                isInitialized = true

                // Pre-load ads after init
                preloadInterstitial(null)
                preloadRewarded(null)

                invoke.resolve()
            }
        }
    }

    @Command
    fun loadInterstitial(invoke: Invoke) {
        val args = try {
            invoke.parseArgs(AdUnitArgs::class.java)
        } catch (e: Exception) {
            null
        }
        preloadInterstitial(args?.adUnitId)
        invoke.resolve()
    }

    @Command
    fun showInterstitial(invoke: Invoke) {
        activity.runOnUiThread {
            val ad = interstitialAd
            if (ad == null) {
                preloadInterstitial(null)
                val result = JSObject()
                result.put("shown", false)
                result.put("rewarded", false)
                result.put("error", "no interstitial loaded")
                invoke.resolve(result)
                return@runOnUiThread
            }

            ad.fullScreenContentCallback = object : FullScreenContentCallback() {
                override fun onAdDismissedFullScreenContent() {
                    interstitialAd = null
                    preloadInterstitial(null)
                    val result = JSObject()
                    result.put("shown", true)
                    result.put("rewarded", false)
                    invoke.resolve(result)
                }

                override fun onAdFailedToShowFullScreenContent(error: com.google.android.gms.ads.AdError) {
                    println("[ADMOB] Failed to show interstitial: ${error.message}")
                    interstitialAd = null
                    preloadInterstitial(null)
                    val result = JSObject()
                    result.put("shown", false)
                    result.put("rewarded", false)
                    result.put("error", error.message)
                    invoke.resolve(result)
                }
            }

            ad.show(activity)
        }
    }

    @Command
    fun loadRewarded(invoke: Invoke) {
        val args = try {
            invoke.parseArgs(AdUnitArgs::class.java)
        } catch (e: Exception) {
            null
        }
        preloadRewarded(args?.adUnitId)
        invoke.resolve()
    }

    @Command
    fun showRewarded(invoke: Invoke) {
        activity.runOnUiThread {
            val ad = rewardedAd
            if (ad == null) {
                preloadRewarded(null)
                val result = JSObject()
                result.put("shown", false)
                result.put("rewarded", false)
                result.put("error", "no rewarded ad loaded")
                invoke.resolve(result)
                return@runOnUiThread
            }

            var userEarnedReward = false

            ad.fullScreenContentCallback = object : FullScreenContentCallback() {
                override fun onAdDismissedFullScreenContent() {
                    rewardedAd = null
                    preloadRewarded(null)
                    val result = JSObject()
                    result.put("shown", true)
                    result.put("rewarded", userEarnedReward)
                    invoke.resolve(result)
                }

                override fun onAdFailedToShowFullScreenContent(error: com.google.android.gms.ads.AdError) {
                    println("[ADMOB] Failed to show rewarded: ${error.message}")
                    rewardedAd = null
                    preloadRewarded(null)
                    val result = JSObject()
                    result.put("shown", false)
                    result.put("rewarded", false)
                    result.put("error", error.message)
                    invoke.resolve(result)
                }
            }

            ad.show(activity) { reward ->
                userEarnedReward = true
                println("[ADMOB] User earned reward: ${reward.amount} ${reward.type}")
            }
        }
    }

    // MARK: - Private

    private fun preloadInterstitial(adUnitId: String?) {
        val unitId = adUnitId ?: testInterstitialId
        val request = AdRequest.Builder().build()

        InterstitialAd.load(activity, unitId, request, object : InterstitialAdLoadCallback() {
            override fun onAdLoaded(ad: InterstitialAd) {
                interstitialAd = ad
                println("[ADMOB] Interstitial loaded")
            }

            override fun onAdFailedToLoad(error: LoadAdError) {
                println("[ADMOB] Interstitial load failed: ${error.message}")
                interstitialAd = null
            }
        })
    }

    private fun preloadRewarded(adUnitId: String?) {
        val unitId = adUnitId ?: testRewardedId
        val request = AdRequest.Builder().build()

        RewardedAd.load(activity, unitId, request, object : RewardedAdLoadCallback() {
            override fun onAdLoaded(ad: RewardedAd) {
                rewardedAd = ad
                println("[ADMOB] Rewarded loaded")
            }

            override fun onAdFailedToLoad(error: LoadAdError) {
                println("[ADMOB] Rewarded load failed: ${error.message}")
                rewardedAd = null
            }
        })
    }
}

File: corpan/plugins/tauri-plugin-admob/ios/Sources/AdmobPlugin.swift
import GoogleMobileAds
import Tauri
import UIKit

class AdmobPlugin: Plugin {
    private var interstitialAd: GADInterstitialAd?
    private var rewardedAd: GADRewardedAd?
    private var isInitialized = false

    // Default test ad unit IDs (Google's official test ads)
    private let testInterstitialId = "ca-app-pub-3940256099942544/4411468910"
    private let testRewardedId = "ca-app-pub-3940256099942544/1712485313"

    override init() {
        super.init()
    }

    // MARK: - Plugin Commands

    @objc func initAdmob(_ invoke: Invoke) throws {
        if isInitialized {
            invoke.resolve()
            return
        }

        DispatchQueue.main.async { [weak self] in
            GADMobileAds.sharedInstance().start { status in
                print("[ADMOB] SDK initialized: \(status.adapterStatusesByClassName)")
                self?.isInitialized = true

                // Pre-load ads after init
                self?.preloadInterstitial(adUnitId: nil)
                self?.preloadRewarded(adUnitId: nil)

                invoke.resolve()
            }
        }
    }

    @objc func loadInterstitial(_ invoke: Invoke) throws {
        let args = try invoke.parseArgs(AdUnitArgs.self)
        preloadInterstitial(adUnitId: args.adUnitId)
        invoke.resolve()
    }

    @objc func showInterstitial(_ invoke: Invoke) throws {
        DispatchQueue.main.async { [weak self] in
            guard let self = self else {
                invoke.resolve(["shown": false, "rewarded": false, "error": "plugin deallocated"])
                return
            }

            guard let ad = self.interstitialAd else {
                // No ad loaded — try loading one for next time
                self.preloadInterstitial(adUnitId: nil)
                invoke.resolve(["shown": false, "rewarded": false, "error": "no interstitial loaded"])
                return
            }

            guard let rootVC = self.getRootViewController() else {
                invoke.resolve(["shown": false, "rewarded": false, "error": "no root view controller"])
                return
            }

            let delegate = InterstitialDelegate { [weak self] in
                // Ad dismissed — preload next one
                self?.interstitialAd = nil
                self?.preloadInterstitial(adUnitId: nil)
                invoke.resolve(["shown": true, "rewarded": false])
            }

            // Store delegate to prevent deallocation
            objc_setAssociatedObject(ad, "delegate", delegate, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
            ad.fullScreenContentDelegate = delegate

            ad.present(fromRootViewController: rootVC)
        }
    }

    @objc func loadRewarded(_ invoke: Invoke) throws {
        let args = try invoke.parseArgs(AdUnitArgs.self)
        preloadRewarded(adUnitId: args.adUnitId)
        invoke.resolve()
    }

    @objc func showRewarded(_ invoke: Invoke) throws {
        DispatchQueue.main.async { [weak self] in
            guard let self = self else {
                invoke.resolve(["shown": false, "rewarded": false, "error": "plugin deallocated"])
                return
            }

            guard let ad = self.rewardedAd else {
                self.preloadRewarded(adUnitId: nil)
                invoke.resolve(["shown": false, "rewarded": false, "error": "no rewarded ad loaded"])
                return
            }

            guard let rootVC = self.getRootViewController() else {
                invoke.resolve(["shown": false, "rewarded": false, "error": "no root view controller"])
                return
            }

            var userEarnedReward = false

            let delegate = RewardedDelegate { [weak self] in
                self?.rewardedAd = nil
                self?.preloadRewarded(adUnitId: nil)
                invoke.resolve(["shown": true, "rewarded": userEarnedReward])
            }

            objc_setAssociatedObject(ad, "delegate", delegate, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
            ad.fullScreenContentDelegate = delegate

            ad.present(fromRootViewController: rootVC) {
                userEarnedReward = true
                print("[ADMOB] User earned reward: \(ad.adReward.amount) \(ad.adReward.type)")
            }
        }
    }

    // MARK: - Private

    private func preloadInterstitial(adUnitId: String?) {
        let unitId = adUnitId ?? testInterstitialId
        GADInterstitialAd.load(withAdUnitID: unitId, request: GADRequest()) { [weak self] ad, error in
            if let error = error {
                print("[ADMOB] Interstitial load failed: \(error.localizedDescription)")
                return
            }
            self?.interstitialAd = ad
            print("[ADMOB] Interstitial loaded")
        }
    }

    private func preloadRewarded(adUnitId: String?) {
        let unitId = adUnitId ?? testRewardedId
        GADRewardedAd.load(withAdUnitID: unitId, request: GADRequest()) { [weak self] ad, error in
            if let error = error {
                print("[ADMOB] Rewarded load failed: \(error.localizedDescription)")
                return
            }
            self?.rewardedAd = ad
            print("[ADMOB] Rewarded loaded")
        }
    }

    private func getRootViewController() -> UIViewController? {
        if #available(iOS 15.0, *) {
            return UIApplication.shared.connectedScenes
                .compactMap { $0 as? UIWindowScene }
                .flatMap { $0.windows }
                .first { $0.isKeyWindow }?
                .rootViewController
        } else {
            return UIApplication.shared.windows.first { $0.isKeyWindow }?.rootViewController
        }
    }
}

// MARK: - Delegates

private class InterstitialDelegate: NSObject, GADFullScreenContentDelegate {
    let onDismiss: () -> Void

    init(onDismiss: @escaping () -> Void) {
        self.onDismiss = onDismiss
    }

    func adDidDismissFullScreenContent(_ ad: GADFullScreenPresentingAd) {
        onDismiss()
    }

    func ad(_ ad: GADFullScreenPresentingAd, didFailToPresentFullScreenContentWithError error: Error) {
        print("[ADMOB] Failed to present: \(error.localizedDescription)")
        onDismiss()
    }
}

private class RewardedDelegate: NSObject, GADFullScreenContentDelegate {
    let onDismiss: () -> Void

    init(onDismiss: @escaping () -> Void) {
        self.onDismiss = onDismiss
    }

    func adDidDismissFullScreenContent(_ ad: GADFullScreenPresentingAd) {
        onDismiss()
    }

    func ad(_ ad: GADFullScreenPresentingAd, didFailToPresentFullScreenContentWithError error: Error) {
        print("[ADMOB] Failed to present rewarded: \(error.localizedDescription)")
        onDismiss()
    }
}

// MARK: - Argument Types

struct AdUnitArgs: Decodable {
    let adUnitId: String?
}

@_cdecl("init_plugin_admob")
func initPlugin() -> Plugin {
    return AdmobPlugin()
}
